API for runtime parameters, e.g. for configuring federation upstream components.

niclic opened the pull request:
### What this PR does:
* Adds additional client methods to allow reading of federation upstream parameters.
* Adds tests for all existing and new federation methods.


### Why do we need this PR?
* To add federation support to the [terraform-provider-rabbitmq](https://www.terraform.io/docs/providers/rabbitmq/index.html) terraform provider, we need to be able to read federation upstream resources via the API. I'm working on this now but need these changes before I can continue.

### Notes
Some notes from working on this PR.


#### API changes
To support the new read methods, I added additional fields to `FederationUpstream` and introduced a `FederationDefinitionDTO` type. This is an exact copy of the existing shovel implementation, which makes sense, because they are more or less the same thing.


#### Default values
The only required field for a federation upstream definition is `Uri`. Although many of the other arguments have stated defaults in the docs, when a value is not provided, it gets the default value for the go type instead. 

For example:

```json
// exported configuration for an upstream with only uri provided
{
  "rabbit_version": "3.8.3",
  "parameters": [
    {
      "value": {
        "exchange": "",
        "expires": 0,
        "max-hops": 0,
        "message-ttl": 0,
        "prefetch-count": 0,
        "queue": "",
        "reconnect-delay": 0,
        "trust-user-id": false,
        "uri": "amqp://127.0.0.1/%2f"
      },
      "vhost": "rabbit/hole",
      "component": "federation-upstream",
      "name": "temporary"
    }
  ]
}
```

Is the value for `prefetch-count` the default of `1000`, or `0`? I don't think these values are the same, although I haven't tested this to be honest. It is easy to supply the default values in a request, which is probably the right thing to do. However, some arguments, such as `expires`, use a default value of _none_ (according to the docs) or _leave blank_ (in the Admin UI) to mean _never expires_. Does `0` accomplish the same thing?

Might it be a good idea to annotate these fields with `omitempty` instead?


#### Deleting a federation upstream parameter
Deleting a non-existent upstream returns a `HTTP 404` response and _not_ an `error`. 

```sh
curl -i -u guest:guest \
-X DELETE http://localhost:15672/api/parameters/federation-upstream/rabbit%2fhole/temporary

HTTP/1.1 404 Not Found
content-length: 49
content-security-policy: script-src 'self' 'unsafe-eval' 'unsafe-inline'; object-src 'self'
content-type: application/json
date: Sun, 26 Apr 2020 14:38:08 GMT
server: Cowboy
vary: accept, accept-encoding, origin

{"error":"Object Not Found","reason":"Not Found"}

```

Same for `PutFederationUpstream`. Both methods use the `executeResponse` function, which does not parse the response into an `ErrorResponse{}` error.

This is probably intentional, but I've added tests for these cases in case this logic is ever changed.


#### Additional federation arguments

* The [federation reference docs](https://www.rabbitmq.com/federation-reference.html#upstreams) show an additional federated queue argument called `consumer-tag`.

* The `Admin > Federation Upstreams > Add a new upstream` UI shows an additional federated exchange argument called `ha-policy`.

* The federation plugin [source](https://github.com/rabbitmq/rabbitmq-federation/blob/master/src/rabbit_federation_parameters.erl#L87) shows some additional arguments that may not be documented yet: `resource-cleanup-mode` and `bind-nowait`.

Perhaps the API should support these additional arguments as well?

michaelklishin commented on federation.go:
```diff
@@ -24,19 +24,83 @@ type FederationDefinition struct {
 
 // Represents a configured Federation upstream.
 type FederationUpstream struct {
```
The tuple of `(vhost, component, name, and value)` is a runtime parameter. An upstream is a specific type (component) of a parameter. So this has to be renamed, e.g. to `RuntimeParameter`.

michaelklishin commented on federation.go:
```diff
@@ -24,19 +24,83 @@ type FederationDefinition struct {
+//
+
+// ListFederationUpstreams returns all federation upstreams
+func (c *Client) ListFederationUpstreams() (rec []FederationUpstream, err error) {
```
It would be very nice to accompany this (and, perhaps, base it on)  with a set of functions that would list, set, and delete arbitrary runtime parameters. Note that RabbitMQ nodes will perform parameter validation so if a component is not known (not registered by a plugin or RabbitMQ core), it will be rejected.

We can also add functions that work with parameters of a certain type. Then federation upstream management would rely on them and pass `federation-upstream` for component. How does this sound?

michaelklishin commented:
Note that the suggested API changes would not require a lot of new tests. An absolute minimum would do, and the existing federation upstream tests would provide most coverage.

niclic commented:
Sounds good. When I first came to adding support for managing federation upstreams to terraform, I expected to be defining parameter types and resources. However, I was swayed by the existence of both Federation and Shovel types already present in rabbithole. I probably should have opened a discussion beforehand. I think adding a set of methods that operate explicitly on runtime parameters aligns well with the underlying RabbitMQ APIs.

The [docs](https://www.rabbitmq.com/parameters.html) describe using _vhost-scoped_ parameters to configure settings at run time. I use your suggested name of `RuntimeParameter`.

A possible API design for managing runtime parameters might look like this. 

This mirrors the underlying HTTP API (except for the order of path params) and follows existing rabbithole API conventions. The corresponding HTTP API calls are shown in the comments.

This is very much influenced by the Federation and Shovel plugins, but perhaps there are other things to consider? There are already methods for managing Shovels and Policies.

```golang

type RuntimeParameter struct {
  Name      string                `json:"name"`
  Vhost     string                `json:"vhost"`
  Component string                `json:"component"`
  Value     RuntimeParameterValue `json:"value"`
}

// could also just use the go type directly or interface{}
type RuntimeParameterValue map[string]interface{}


// /api/parameters A list of all vhost-scoped parameters.
ListRuntimeParameters()
  => []RuntimeParameter, error

// /api/parameters/component   A list of all vhost-scoped parameters for a given component.
ListRuntimeParametersFor(component string)
  => []RuntimeParameter, error

ListRuntimeParametersIn(vhost, component string)
  => []RuntimeParameter, error

// /api/parameters/component/vhost/name  An individual vhost-scoped parameter.
GetRuntimeParameter(vhost, component, name string)
  => *RuntimeParameter, error

PutRuntimeParameter(vhost, component, name string, value RuntimeParameterValue)
  => error

DeleteRuntimeParameter(vhost, component, name string)
  => error

```

Thoughts?

niclic commented:
The above are meant to be public methods, but if there are also methods for managing federation or other "types" of runtime parameters, then perhaps a separate set of exported methods for managing generic parameters will be redundant or confusing.

Perhaps this is what you were getting at. To create a set of methods and/or functions, probably private, that manage interacting with the underlying parameter API.

Certainly, some private functions will be required to handle the marshalling of specific types to and from json and avoid the overhead of writing conversion code.

For example.

```golang
// federation.go
func (c *Client) GetFederationUpstream(vhost, upstreamName string) (upstream *FederationUpstream, err error) {
    return getRuntimeParameter(vhost, "federation-upstream", &upstream)
}

// runtime_parameter.go
func getRuntimeParameter(vhost, name string, rec interface{}) err error {
  // ...
  executeAndParseRequest(c, req, &rec)
  // ...
}

```

michaelklishin commented:
The proposed API makes sense to me. Additional public functions that operate specifically on upstreams also make sense to me. Your example with `GetFederationUpstream` is indeed what I had in mind but `getRuntimeParameter` can be `GetRuntimeParameter` instead (so, public). Managing runtime parameters is a thing that operators might want to do and automate :)

michaelklishin commented:
Thank you for your willingness to invest more time into this PR!

niclic commented:
Okay, please review these changes.

This is still WIP, but I wanted to check in before moving completing the changes.

#### Changes so far

1. Added an API for runtime (vhost-scoped) parameters. See this [commit](https://github.com/michaelklishin/rabbit-hole/pull/150/commits/bbad1c4f88f00a26350d3d868e3688b0be1933ce).

This allows you to work with parameter values containing arbitrary data. See this [test](https://github.com/michaelklishin/rabbit-hole/pull/150/commits/bbad1c4f88f00a26350d3d868e3688b0be1933ce#diff-2ec841ccc848feb12b96dfcd1e33326fR2176).
You can use this API to manage runtime parameters directly and forget about the other APIs. But you will have to work directly with `map[string]interface{}`.

2. Added some V2 methods to `federation.go` for `PUT`ting and `GET`ting runtime parameters that have a component of `federation-upstream`.



I have only added a couple of temporary V2 methods to show one approach to using the runtime parameter API to manage different types of runtime parameters.

* [PutFederationUpstreamV2](https://github.com/michaelklishin/rabbit-hole/pull/150/commits/5f31f0ce25990d382cf4b9c84a76c9938095937b#diff-7760fa4beb711b6df4dc1973473f022bR120)
* [GetFederationUpstreamV2](https://github.com/michaelklishin/rabbit-hole/pull/150/commits/84c7b101eb6107f5dfa037bbfb6bcf0f4294ac81#diff-7760fa4beb711b6df4dc1973473f022bR102)

3. To avoid any naming confusion, I renamed `FederationUpstream` to `FederationInfo`. 

This approach requires a `struct` to represent a strongly-typed parameter value and this mirrors the approach used for Shovels. Note the [PopulateRuntimeParameter](https://github.com/michaelklishin/rabbit-hole/pull/150/commits/84c7b101eb6107f5dfa037bbfb6bcf0f4294ac81#diff-934a493982c565232fb70bc1bd55785aR89) method. We need to pass the desired `struct` to the JSON decoder. Otherwise, you will have to do some form of mapping. This method was private to begin with, but I made it public.


#### Other approaches worth considering.

1. The parameter API could just return `RuntimeParameter` and the federation API would be responsble for mapping from `map[string]interface{}` to `FederationInfo`. 

This is a common approach, but requries extra code to safely access map values. This would remove the need for the `PopulateRuntimeParameter` method. So, the API would be more consistent.

This sort of thing.

```golang
  // inside GetFederationUpstream
  param, err := GetRuntimeParameter()

  info := &FederationInfo{
    Name: param.Name,
    Definition: FederationDefinition{} 
  }

  m := param.Value.(map[string]interface{})
  if v, ok := m["uri"].(string); ok {
    info.Definition.Uri = v
  }

  return info
  // etc

```
2. I don't think returning `RuntimeParameter` from the federation API itself is an option, since the end user would need to do some some of conversion.

3. However, you _could_ embed `RuntimeParamter` in `FederationInfo`. This would help to cement the relationship between federation and runtime parameters, but is really only a semantic change. This uses field shadowing to ignore the `RuntimeParameter.Value` field.

```golang
type FederationInfo struct {
  RuntimeParameter
  Value FederationDefinition `json:"value"`
}
// decodes to the following json
// {
//   "name": "temporary",
//   "vhost": "rabbit/hole",
//   "component": "federation-upstream",
//   "value": {
//     "uri": "amqp://127.0.0.1/%2f",
//     "expires": 0,
//     "message-ttl": 0,
//     "max-hops": 0,
//     "prefetch-count": 1000,
//     "reconnect-delay": 1,
//     "ack-mode": "on-confirm",
//     "trust-user-id": false,
//     "exchange": "",
//     "queue": ""
//   }
// }
```

There is probably a more elegant solution to all of this, but this seems pretty consistent with the overall API now.

Thoughts?

michaelklishin commented:
This looks close to what I had in mind. Constructing a federation upstream (info if you will) struct from a runtime parameter struct makes most sense to me. Embedding the parameter feels like a loss of "encapsulation" for no particular benefit.

I personally think that `FederationUpstream` is the best name. "Federation info" is not very specific since there are like three "parts" that make up a federated entity: an upstream, a link, and the entity itself (some may even add a policy), and the struct clearly only covers the first part.

Thanks 🙏

michaelklishin commented:
Thank you so much! This is a great addition to the library.
